### DISCUSSION TITLE
`HashMap<String, DisplayFromStr>`

### DISCUSSION THREAD

**Discussion by slinkydeveloper** [jonasbb/serde_with#473] [Q&A] [answered]
Hi,
I'm trying to use serde_as to deserialize (without serializing) `HashMap<String, http::Uri>`, where `http::Uri` implements `FromStr`. My struct looks like the following (stripped down of all the other fields):

```rust
#[serde_as]
#[derive(Debug, Clone, Deserialize)]
pub struct Configuration {
    #[serde_as(as = "HashMap<String, DisplayFromStr>")]
    pub endpoints: HashMap<String, Uri>,
}
```

But I can't make this working, I get this compilation error:

```
error[E0277]: the trait bound `HashMap<std::string::String, DisplayFromStr>: DeserializeAs<'_, _>` is not satisfied
   --> src/config/mod.rs:13:1
    |
13  |   #[serde_as]
    |   ^----------
    |   |
    |  _in this procedural macro expansion
    | |
14  | | #[derive(Debug, Clone, Deserialize, Builder)]
    | |__________________________________^ the trait `DeserializeAs<'_, _>` is not implemented for `HashMap<std::string::String, DisplayFromStr>`
    |
    = help: the following implementations were found:
              <HashMap<KAs, VAs> as DeserializeAs<'de, BTreeSet<(K, V)>>>
              <HashMap<KAs, VAs> as DeserializeAs<'de, BinaryHeap<(K, V)>>>
              <HashMap<KAs, VAs> as DeserializeAs<'de, HashSet<(K, V)>>>
              <HashMap<KAs, VAs> as DeserializeAs<'de, LinkedList<(K, V)>>>
            and 5 others
note: required by a bound in `As::<T>::deserialize`
   --> /home/slinkydeveloper/.cargo/registry/src/github.com-1ecc6299db9ec823/serde_with-1.13.0/src/lib.rs:428:12
    |
428 |         T: DeserializeAs<'de, I>,
    |            ^^^^^^^^^^^^^^^^^^^^^ required by this bound in `As::<T>::deserialize`
    = note: this error originates in the attribute macro `serde_as` (in Nightly builds, run with -Z macro-backtrace for more info)

For more information about this error, try `rustc --explain E0277`.

```

Interestingly, `Option<Uri>` worked fine (by using as `Option<DisplayFromStr>`). Any ideas?

Thank you for your help!

**Comment by slinkydeveloper** [comment, marked as answer]
I've found out the solution :smile: 

```rust
#[serde_as(as = "HashMap<serde_with::Same, DisplayFromStr>")]
```

The reason is that the implementation of `DeserializeAs` for `HashMap` requires both K and V to implement `DeserializeAs` (see https://docs.rs/serde_with/1.14.0/serde_with/de/trait.DeserializeAs.html#impl-DeserializeAs%3C%27de%2C%20HashMap%3CK%2C%20V%2C%20S%3E%3E-for-HashMap%3CKU%2C%20VU%2C%20S%3E), but apparently `String` is not automatically inferred to implement `DeserializeAs`, hence using [`Same`](https://docs.rs/serde_with/1.14.0/serde_with/struct.Same.html) forces `String` to be resolved as `DeserializeAs` through its serde `Deserialize` implementation. Also using:

```rust
#[serde_as(as = "HashMap<_, DisplayFromStr>")]
```

works fine.

**Reply by jonasbb** [reply]
Glad you found the solution :) The recommended way is using `_` whenever you want to use the normal `Serialize`/`Deserialize` implementations of a type (here serialize a `String` with `<String as Serialize>::serialize`. The examples/documentation all use this. The issue is #202. I am sadly not aware of any way to fix this or even to generate better error messages.

**Reply by slinkydeveloper** [reply]
I think it's already easy enough, but yep i agree it would be better to have a less cryptic error message in this case :)
